File: Python/AddTwoNumbersLL.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def add_two_numbers(l1: ListNode, l2: ListNode) -> ListNode:
    dummy = ListNode(0)
    tail = dummy
    carry = 0

    while l1 or l2 or carry:
        total = carry
        if l1:
            total += l1.val
            l1 = l1.next
        if l2:
            total += l2.val
            l2 = l2.next

        carry = total // 10
        tail.next = ListNode(total % 10)
        tail = tail.next

    return dummy.next

# Helper to print list
def print_list(head: ListNode):
    while head:
        print(head.val, end=" -> ")
        head = head.next
    print("None")

# Example 1
l1 = ListNode(5, ListNode(4))
l2 = ListNode(4)
result1 = add_two_numbers(l1, l2)
print_list(result1)  # Output: 9 -> 4 -> None

# Example 2
l3 = ListNode(4, ListNode(5, ListNode(6)))
l4 = ListNode(1, ListNode(2, ListNode(3)))
result2 = add_two_numbers(l3, l4)
print_list(result2)  # Output: 5 -> 7 -> 9 -> None

File: JAVA/AddTwoNumbersLL.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; this.next = null; }
}

public class AddTwoNumbersLL {

    public static ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        ListNode tail = dummy;
        int carry = 0;

        while (l1 != null || l2 != null || carry != 0) {
            int sum = carry;
            if (l1 != null) {
                sum += l1.val;
                l1 = l1.next;
            }
            if (l2 != null) {
                sum += l2.val;
                l2 = l2.next;
            }

            carry = sum / 10;
            tail.next = new ListNode(sum % 10);
            tail = tail.next;
        }

        return dummy.next;
    }

    // Helper to print list
    public static void printList(ListNode head) {
        while (head != null) {
            System.out.print(head.val + " -> ");
            head = head.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        // Example 1: l1 = 5->4, l2 = 4
        ListNode l1 = new ListNode(5);
        l1.next = new ListNode(4);
        ListNode l2 = new ListNode(4);
        ListNode result1 = addTwoNumbers(l1, l2);
        printList(result1); // Output: 9 -> 4 -> null

        // Example 2: l1 = 4->5->6, l2 = 1->2->3
        ListNode l3 = new ListNode(4);
        l3.next = new ListNode(5);
        l3.next.next = new ListNode(6);
        ListNode l4 = new ListNode(1);
        l4.next = new ListNode(2);
        l4.next.next = new ListNode(3);
        ListNode result2 = addTwoNumbers(l3, l4);
        printList(result2); // Output: 5 -> 7 -> 9 -> null
    }
}
